lint/src/main/java/leakcanary/lint/ProductionUsageDetector.kt

```diff
@@ -0,0 +1,53 @@
+package leakcanary.lint
+
+import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.GradleContext
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity.ERROR
+import leakcanary.lint.ProductionUsageDetector.Companion.DETAIL_MESSAGE
+import leakcanary.lint.ProductionUsageDetector.Companion.ISSUE_ID
+import leakcanary.lint.ProductionUsageDetector.Companion.MESSAGE
+
+val ISSUE_PROD_USAGE_PATTERN = Issue.create(
+    ISSUE_ID,
+    MESSAGE,
+    DETAIL_MESSAGE,
+    CORRECTNESS, 10, ERROR,
+    Implementation(ProductionUsageDetector::class.java, Scope.GRADLE_SCOPE)
+)
+
+class ProductionUsageDetector : Detector(), Detector.GradleScanner {
+
+  companion object {
+    const val BLOCK_NAME = "dependencies"
+    const val DEPENDENCY_PREFIX = "com.squareup.leakcanary"
+    const val RELEASE_ENVIRONMENT = "release"
+    const val IMPLEMENTATION = "implementation"
+    const val API = "api"
+    const val MESSAGE = "LeakCanary should not be used in production."
+    const val DETAIL_MESSAGE = "$MESSAGE Please use debugImplementation instead."
+    const val ISSUE_ID = "ProductionUsagePattern"
+  }
+
+  override fun checkDslPropertyAssignment(
+    context: GradleContext,
+    property: String,
+    value: String,
+    parent: String,
+    parentParent: String?,
+    propertyCookie: Any,
+    valueCookie: Any,
+    statementCookie: Any
+  ) {
+    if (parent == BLOCK_NAME && value.contains(DEPENDENCY_PREFIX)) {
+      if (property == IMPLEMENTATION || property == API || property.contains(RELEASE_ENVIRONMENT)) {
+        context.report(
+            ISSUE_PROD_USAGE_PATTERN, context.getLocation(propertyCookie), DETAIL_MESSAGE
+        )
```

Comment:
This lint check will likely yield false positives, e.g. we have an internal dev only leakcanary module in our app and that one has an `implementation` dependency on LeakCanary, and then that module is imported with debugImplementation. 

Similarly, I expect that the mistake we saw happen because it was a large codebase, and adding "debugImplementation" wasn't working for whatever set of reasons. In large codebase, you'll rarely see dependencies declared directly, most likely it'll be something like `implementation deps.leakcanary` instead.

I'm sorry to question all this, I know it's a ton of work. I wish there was a way to check against the resolved dependencies for a given configuration, and provide that check automatically as the dependency is added.

Reply:
Lint is working on exposing granular information on the resolved graph, actually. It's needed for checks like https://issuetracker.google.com/issues/77234248 and https://issuetracker.google.com/issues/77246430.

In general Lint favors checks that produce false positives such that you can suppress rather than false negatives. Though I agree it's obviously nice to dramatically minimize those false positives.